### ISSUE TITLE
Replace CDAT

### ISSUE THREAD

**Issue by forsyth2** [E3SM-Project/zppy#346]
https://github.com/conda-forge/cdtime-feedstock/pull/52 mentions that we need to replace `CDAT` in `zppy`. Note that we already have #80, which is specifically about `cdscan`.

**Comment by forsyth2** [comment]
@tomvothecoder I'm not quite sure what needs to be done for `zppy`.
```
$ git grep -n cdscan
docs/source/dev_guide/new_diags_set.rst:100:        cdscan -x ${xml_name} ${rofDir}/${v}_*.nc
zppy/templates/e3sm_diags.bash:104:        # Add this time series file to the list of files for cdscan to use
zppy/templates/e3sm_diags.bash:111:    cdscan -x ${xml_name} -f ${v}_files.txt
zppy/templates/e3sm_diags.bash:132:  cdscan -x ${xml_name} ${ts_rof_dir_source}/${v}_*.nc
zppy/templates/global_time_series.bash:46:cdscan -x glb.xml *.nc
zppy/templates/global_time_series.bash:70:    cdscan -x glb.xml mpaso.glb*.nc

$ git grep -n cdat
# No matches

$ git grep -n cdms
zppy/templates/readTS.py:1:import cdms2
zppy/templates/readTS.py:10:        self.f = cdms2.open(filename)

$ emacs conda/dev.yml
# No dependencies appear obviously related to cdat

$ emacs conda/meta.yml
# No dependencies appear obviously related to cdat
```

**Comment by tomvothecoder** [comment]
@forsyth2 I don't have familiarity with the `zppy` codebase to give input on what needs to be done specifically.

My general approach to refactoring is to identify what needs to be refactored, understand the implementation, and find alternative solutions that would reproduce the same/similar results.

**Comment by tomvothecoder** [comment]
It looks like you identified the CDAT modules that are used in `zppy`, which is a good first step. Luckily, the dependency is not that significant from what I can tell.

**Comment by forsyth2** [comment]
Ok, great!

https://cdat.llnl.gov/ lists "specific CDAT packages (`cdms2`, `cdat_info`, `cdutil`, `cdtime`, `genutil`, `libcdms`)". Out of those, it looks like zppy has:
```
$ git grep -n cdms2
zppy/templates/readTS.py:1:import cdms2
zppy/templates/readTS.py:10:        self.f = cdms2.open(filename)

$ git grep -n cdutil
zppy/templates/readTS.py:2:import cdutil
zppy/templates/readTS.py:65:            v = cdutil.YEAR(v)
```

So, basically I just need replacements for `cdcscan`, `cdms2.open`, and `cdutil.YEAR`.

**Comment by tomvothecoder** [comment]
Those are the packages who's dependencies are being maintained until end of CY2023.

I doubt `zppy` uses other modules from CDAT. It might be a good idea to check other modules that operate on `cdms2` objects, like `MV2`.
* Example from `e3sm_diags`:  https://github.com/E3SM-Project/e3sm_diags/discussions/607
* Complete list of CDAT modules: https://anaconda.org/cdat/repo

**Comment by forsyth2** [comment]
Thanks, out of the list at https://anaconda.org/cdat/repo, we just have:
```
$ git grep -n cdms2
zppy/templates/readTS.py:1:import cdms2
zppy/templates/readTS.py:10:        self.f = cdms2.open(filename)

$ git grep -n cdutil
zppy/templates/readTS.py:2:import cdutil
zppy/templates/readTS.py:65:            v = cdutil.YEAR(v)
```

**Comment by forsyth2** [comment]
We can probably use `xarray` to replace `cdms2.open`, and an XML reader to replace `cdscan` (`mfopen` to open multiple files). May have to replace with multiple tools.

**Comment by forsyth2** [comment]
@tomvothecoder What's the timeline/deadline for this issue? Are we trying not to have anything CDAT-related by the next Unified release?

**Comment by tomvothecoder** [comment]
@forsyth2 Just CC'd you in an email chain with Xylar. He gave us 6-9 months to refactor any E3SM packages from CDAT to alternative solutions so they can work in E3SM Unified.

It looks like `zppy`'s CDAT dependency is minimal and much more straightforward to refactor compared to `e3sm_diags`.

**Comment by forsyth2** [comment]
@tomvothecoder Sounds good; I've moved this issue up in priority, so it can be completed by the next release (February/March of 2024)

**Comment by forsyth2** [comment]
@tomvothecoder I apologize if this is a naive question, but are there `xCDAT` equivalents for all the CDAT commands? That seems like it would most simplify this change.

It looks like we need to change 3 things here:

- Replace `cdms2.open`; I'm assuming this would be done with `xarray.open_dataset` (https://docs.xarray.dev/en/stable/generated/xarray.open_dataset.html) or perhaps `xcdat.open_dataset` (https://xcdat.readthedocs.io/en/stable/generated/xcdat.open_dataset.html#xcdat.open_dataset).
- Replace `cdscan -x`; is there a recommended XML reader that could handle this?
- Replace `cdutil.YEAR`; this seems trickiest. Looking at the file where it's defined, https://github.com/CDAT/cdutil/blob/master/cdutil/times.py, there is quite a bit of functionality. Would the best option be to determine which functionality is needed by `zppy` and write that myself??

**Comment by tomvothecoder** [comment]
>  Replace `cdms2.open`; I'm assuming this would be done with `xarray.open_dataset` ([docs.xarray.dev/en/stable/generated/xarray.open_dataset.html](https://docs.xarray.dev/en/stable/generated/xarray.open_dataset.html)) or perhaps `xcdat.open_dataset` ([xcdat.readthedocs.io/en/stable/generated/xcdat.open_dataset.html#xcdat.open_dataset](https://xcdat.readthedocs.io/en/stable/generated/xcdat.open_dataset.html#xcdat.open_dataset)).

Yes, you can use either API if a single dataset is being opened. `xcdat.open_dataset()` has some additional capabilities listed in the docs. If you don't need them, `xr.open_dataset()` will work fine.

> Replace `cdscan -x`; is there a recommended XML reader that could handle this?

In #80, Jill says _"`cdscan` is used to concatenate files before running e3sm-diags and global mean time-series. One possibility is to use xarray's IO to read-in multiple files."_

Since you'll be moving to xarray/xCDAT, I suggest exploring [xarray.open_mfdataset()](https://docs.xarray.dev/en/stable/generated/xarray.open_mfdataset.html) or [xcdat.open_mfdataset()](https://xcdat.readthedocs.io/en/latest/generated/xcdat.open_mfdataset.html). These APIs can read a glob or list of `.nc` file paths and combine them into a single `xr.Dataset`.

> Replace `cdutil.YEAR`; this seems trickiest. Looking at the file where it's defined, [CDAT/cdutil@`master`/cdutil/times.py](https://github.com/CDAT/cdutil/blob/master/cdutil/times.py?rgh-link-date=2023-10-02T23%3A59%3A53Z), there is quite a bit of functionality. Would the best option be to determine which functionality is needed by `zppy` and write that myself??

xCDAT has APIs for temporal averaging. The equivalent to `cdutil.YEAR` should be the xCDAT accessor method, [xr.Dataset.temporal.group_average(...freq="year")](https://xcdat.readthedocs.io/en/stable/generated/xr.Dataset.temporal.average.html#xarray.Dataset.temporal.group_average). The docs include examples on how to use it.

---

Here's a CDAT to xCDAT mapping table (not all features are 1:1): https://xcdat.readthedocs.io/en/stable/api.html#cdat-mapping-table

**Comment by forsyth2** [comment]
Reopening since #519 was a partial resolution.
